### PR TITLE
AAD update

### PR DESCRIPTION
**Pull request by mrm9084**
Update to the latest SDK version, adds AAD Support. Adds support for TokenCredentialProvier so users can provider Azure TokenCredentials to App Configuration.

Note: Draft as current SDK isn't released yet, but needs review.

### CODE REVIEW

File: spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md [hunk 1]
```diff
@@ -37,77 +40,126 @@ spring.cloud.azure.appconfiguration.fail-fast | Whether throw RuntimeException o
...
 Failfast feature decides whether throw RuntimeException or not when exception happens. By default, failfast is enabled, it can be disabled with below configuration:
-```
+
+```properties
 spring.cloud.azure.appconfiguration.fail-fast=false
 ```
 
-### Use Azure Managed Identity to load the connection string
-[Managed service identity](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview) allows application to access [Azure Active Directory](https://azure.microsoft.com/en-us/services/active-directory/) protected resource on [Azure](https://azure.microsoft.com/en-us/).
+### Use Azure Managed Identity/Service Principle to load the connection string
```
Comment (zhenlan) [inline spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md]: I find combining managed identity and service principle is very confusing. We should separate them. For example, you can have two sections
#### Use Managed Identity to access App Configuration
#### Use Azure Active Directory to access App Configuration

There is nothing about loading connection string here. It should be removed.

File: spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md [hunk 2]
```diff
@@ -37,77 +40,126 @@ spring.cloud.azure.appconfiguration.fail-fast | Whether throw RuntimeException o
...
+
+[Managed service identity](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview) allows application to access [Azure Active Directory](https://azure.microsoft.com/services/active-directory/) protected resource on [Azure](https://azure.microsoft.com).
```
Comment (zhenlan) [inline spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md]: Managed ~service~ identity

File: spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md [hunk 3]
```diff
@@ -37,77 +40,126 @@ spring.cloud.azure.appconfiguration.fail-fast | Whether throw RuntimeException o
...
 
-In this library, managed service identity is used to retrieve the connection string of the configuration store, the connection string is not required if running the Spring Boot application on Azure with managed service identity enabled.
+In this library, [Azure Identity SDK][azure_identity_sdk] is used to access Azure App Configuration and optionally Azure Key Vault, for secrets. The connection string is not required and will be ignored if Managed Identity is being used.
 
 Follow below steps to enable managed service identity feature:
```
Comment (zhenlan) [inline spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md]: Follow below steps to enable *accessing App Configuration with* managed ~service~ identities ~feature~:

File: spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md [hunk 4]
```diff
@@ -37,77 +40,126 @@ spring.cloud.azure.appconfiguration.fail-fast | Whether throw RuntimeException o
...
 
-1. [Enable managed identities service](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview#how-can-i-use-managed-identities-for-azure-resources) for virtual machine or App Service, on which the application will be deployed
+1. [Enable managed identities service](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview#how-can-i-use-managed-identities-for-azure-resources) for virtual machine or App Service, on which the application will be deployed
```
Comment (zhenlan) [inline spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md]: It doesn't have to be only VM or App Service. For example,

[Enable managed identities](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview#how-can-i-use-managed-identities-for-azure-resources) for the [supported Azure services](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/services-support-managed-identities), for example, virtual machine or App Service, on which the application is deployed

File: spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md [hunk 5]
```diff
@@ -37,77 +40,126 @@ spring.cloud.azure.appconfiguration.fail-fast | Whether throw RuntimeException o
...
+
+1. Configure the [Azure RBAC](https://docs.microsoft.com/azure/role-based-access-control/role-assignments-portal) to allow application running on VM or App Service to access the configuration store. When prompted for the Role select App Configuration Data Reader, App Configuration Data Owner is not required but can be used if needed.
```
Comment (zhenlan) [inline spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md]: Configure the [Azure RBAC](https://docs.microsoft.com/azure/role-based-access-control/role-assignments-portal) of your App Configuration store to grant access to the Azure service where your application is running. Select the *App Configuration Data Reader* role. The *App Configuration Data Owner* role is not required but can be used if needed.

File: spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md [hunk 6]
```diff
@@ -37,77 +40,126 @@ spring.cloud.azure.appconfiguration.fail-fast | Whether throw RuntimeException o
...
+
+1. Choose Configuration option:
```
Comment (zhenlan) [inline spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md]: I'm not sure I'm following. Let's discuss offline.

File: spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md [hunk 7]
```diff
@@ -37,77 +40,126 @@ spring.cloud.azure.appconfiguration.fail-fast | Whether throw RuntimeException o
...
+                    .build();
+    }
+
+}
 ```
+
+### bootstrap.application
+
+```application
 spring.cloud.azure.appconfiguration.stores[0].name=[config-store-name]
```
Comment (zhenlan) [inline spring-cloud-azure-starters/spring-cloud-starter-azure-appconfiguration-config/README.md]: I think we will want to deprecate this and replace it with `spring.cloud.azure.appconfiguration.stores[0].endpoint`

**Comment by neuqlz** [comment]
@mrm9084 hi CI seems failed. Could you fix it?

**Comment by mrm9084** [comment]
Hi, that is known because the sdk isn’t released yet and wanted the changes reviews early. He we’re hoping you could start the review with it in this state. There is a note in my pull request description.

The sdk should be out today so I will update it then.

Thanks,
Matthew Metcalf

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java [hunk 8]
```diff
@@ -175,17 +183,16 @@ private String getKeyVaultEntry(String value) {
...
-            Duration keyVaultWaitTime = Duration.ofSeconds(appProperties.getKeyVaultWaitTime());
-            Secret secret = keyVaultClients.get(uri.getHost()).getSecret(uri, keyVaultWaitTime);
+            KeyVaultSecret secret = keyVaultClients.get(uri.getHost()).getSecret(uri, appProperties.getMaxRetryTime());
             if (secret == null) {
                 throw new IOException("No Key Vault Secret found for Reference.");
             }
             secretValue = secret.getValue();
         } catch (RuntimeException | IOException e) {
-            if (azureProperties.isFailFast()) {
+            if (!azureProperties.isFailFast()) {
```
Comment (x9xu) [inline spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java]: Wondering why the logic is flipped here? if fail fast, we are supposed to throw exception here?

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java [hunk 9]
```diff
@@ -40,80 +33,68 @@
...
+        TokenCredentialProvider tokenCredentialProvider = null;
+        try {
+            tokenCredentialProvider = context.getBean(TokenCredentialProvider.class);
+        } catch (NoUniqueBeanDefinitionException e) {
+            LOGGER.error("Failed to find unique TokenCredentialProvider Bean for authentication.", e);
+            if (properties.isFailFast()) {
+                throw e;
+            }
+        } catch (NoSuchBeanDefinitionException e) {
+            LOGGER.info("No TokenCredentialProvider found.");
```
Comment (x9xu) [inline spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java]: Wondering if we need to make sure that ClientStore always has a TokenCredentialProvider here?

File: spring-cloud-azure-appconfiguration-config/src/test/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceKeyVaultTest.java [hunk 10]
```diff
@@ -144,8 +145,8 @@ public void testKeyVaultTest() throws Exception {
         KeyVaultClient client = Mockito.mock(KeyVaultClient.class);
         PowerMockito.whenNew(KeyVaultClient.class).withAnyArguments().thenReturn(client);
 
-        Secret secret = new Secret("mySecret", "mySecret");
-        given(client.getSecret(Mockito.any(URI.class), Mockito.any(Duration.class))).willReturn(secret);
+        KeyVaultSecret secret = new KeyVaultSecret("mySecret", "mySecret");
```
Comment (x9xu) [inline spring-cloud-azure-appconfiguration-config/src/test/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySourceKeyVaultTest.java]: Suggest to use different values here for clear demo

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java (hunk 9 above)
Comment (mrm9084) [reply spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java]: They don't need to specify a TokenCredentialProvider, an example is a connection-string being specified.

File: spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java (hunk 8 above)
Comment (mrm9084) [reply spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigPropertySource.java]: It was incorect in the previus release resulting in fail-fast doing the wrong thing.

**Comment by mrm9084** [comment]
@neuqlz The SDK isn't out yet, so its built with beta libaries. We were hoping for the review to start before hand, this is why it is a draft and not ready for review.

**Comment by saragluna** [comment]
@mrm9084 Hi, Matthew. We've just made our 1.2.0 release to align with Spring Cloud Hoxton, which makes the version of the master branch is now 1.2.x, so could you please make another PR targeting the 1.1.x branch since this feature is planned to go out in 1.1.0.M6? And you may keep this PR if this feature is needed in our 1.2.x or close it otherwise. Sorry for the inconvenience.

**Pull request merged**